lib/metag_utils.pm

```diff
@@ -0,0 +1,1795 @@
+#
+#         -a:  Write protein translations to the selected file.
+#         -c:  Closed ends.  Do not allow genes to run off edges.
+#         -d:  Write nucleotide sequences of genes to the selected file.
+#         -f:  Select output format (gbk, gff, or sco).  Default is gbk.
+#         -g:  Specify a translation table to use (default 11).
+#         -h:  Print help menu and exit.
+#         -i:  Specify FASTA/Genbank input file (default reads from stdin).
+#         -m:  Treat runs of N as masked sequence; don't build genes across them.
+#         -n:  Bypass Shine-Dalgarno trainer and force a full motif scan.
+#         -o:  Specify output file (default writes to stdout).
+#         -p:  Select procedure (single or meta).  Default is single.
+#         -q:  Run quietly (suppress normal stderr output).
+#         -s:  Write all potential genes (with scores) to the selected file.
+#         -t:  Write a training file (if none exists); otherwise, read and use
+#              the specified training file.
+#         -v:  Print version number and exit.
+#--------------------------------------------------------------------------------
+sub _build_prodigal_cmd {
+    my ($self, $input_file, $trans_file, $nuc_file, $output_file, $output_type,
+        $mode, $start_file, $training_file) = @_;
+
+    if (!defined($input_file)) {
+       croak "An input FASTA/Genbank file is required for Prodigal to run.\n";
+    }
+    # setting defaults
+    if (!defined($output_type) || $output_type eq '') {
+        $output_type = 'gff';
+    }
+    if (!defined($mode) || $mode eq '') {
+        $mode = 'meta';
+    }
+    my($vol, $f_path, $file) = splitpath($input_file);
+    if (!defined($output_file) || $output_file eq '') {
+        $output_file = catfile($f_path, 'prodigal_output.'.$output_type);
+    }
+    if (!defined($trans_file) || $trans_file eq '') {
+       $trans_file = catfile($f_path, 'protein_translation');
+    }
+    if (!defined($nuc_file) || $nuc_file eq '') {
+       $nuc_file = catfile($f_path, 'nucleotide_seq');
+    }
+
+    # building the Prodigal command
+    my @cmd = ('/kb/runtime/bin/prodigal');
+    push @cmd, '-i'; push @cmd, "$input_file";
+    push @cmd, '-f'; push @cmd, "$output_type";
+    push @cmd, '-o'; push @cmd, "$output_file";
+    push @cmd, '-a'; push @cmd, "$trans_file";
+    push @cmd, '-d'; push @cmd, "$nuc_file";
+    push @cmd, '-c';
+    push @cmd, '-g'; push @cmd, 11;
+    push @cmd, '-q';
+    push @cmd, '-p'; push @cmd, "$mode";
+    push @cmd, '-m';
+    if ($start_file) {
+        push @cmd, '-s';
+        push @cmd, "$start_file";
+    }
+    if ($training_file) {
+        push @cmd, '-t';
+        push @cmd, "$training_file";
+    }
+    return @cmd;
+}
+
+
+sub _run_prodigal {
+    my ($self, @cmd) = @_;
+    my $ret = 0;
+    eval {
+        $ret = system(@cmd);
+    };
+    if ($@) {
+        print Dumper(\@cmd);
+        croak "ERROR Prodigal run failed: ".$@."\n";
+    }
```

Comment:
For new code, I suggest using `Try::Tiny` instead of `eval` -- see https://metacpan.org/pod/Try::Tiny . There is a lot of problematic, potentially buggy use of `eval` in KBase perl code, so better not to add to it!